File: evernote_client/service.py
"""Shared business logic layer for MCP and CLI."""

from __future__ import annotations

from collections.abc import Callable
from typing import Any

from evernote_client.auth import OAuthError, get_token
from evernote_client.client import EvernoteClient
from evernote_client.config import settings
from evernote_client.models import (
    CreatedNote,
    NotebookInfo,
    NoteContent,
    NoteMetadata,
    SearchResult,
    TagInfo,
)


class PrivateNoteError(PermissionError):
    """Raised when an operation targets a note tagged 'private'."""


_client: EvernoteClient | None = None


def get_client() -> EvernoteClient:
    """Get or create the singleton EvernoteClient."""
    global _client  # noqa: PLW0603
    if _client is None:
        try:
            token = get_token(settings)
        except OAuthError:
            raise
        _client = EvernoteClient(token)
    return _client


def _is_private(tag_guids: list[str]) -> bool:
    guid = get_client().private_tag_guid
    return guid is not None and guid in tag_guids


def resolve_notebook_guid(client: EvernoteClient, name: str) -> str:
    """Resolve notebook name to GUID.

    Raises:
        ValueError: If no notebook matches the given name.
    """
    notebooks = client.list_notebooks()
    for nb in notebooks:
        if nb.name == name and nb.guid is not None:
            return nb.guid
    msg = f"Notebook not found: {name}"
    raise ValueError(msg)


# --- Read operations ---


def search_notes(
    query: str = "",
    notebook_name: str = "",
    tags: list[str] | None = None,
    max_results: int = 20,
    offset: int = 0,
) -> SearchResult:
    """Search notes using Evernote search grammar."""
    client = get_client()
    max_results = min(max_results, 100)

    notebook_guid = None
    if notebook_name:
        notebook_guid = resolve_notebook_guid(client, notebook_name)

    # Strip "private" from tag filters to prevent direct targeting
    safe_tags = ([t for t in tags if t.lower() != "private"] if tags else tags) or None

    result = client.search_notes(
        query=query,
        notebook_guid=notebook_guid,
        tag_names=safe_tags,
        max_results=max_results,
        offset=offset,
    )

    all_notes = [NoteMetadata.from_thrift(n) for n in (result.notes or [])]
    notes = [n for n in all_notes if not _is_private(n.tag_guids)]
    filtered = len(all_notes) - len(notes)

    return SearchResult(
        notes=notes,
        total=max(0, (result.totalNotes or 0) - filtered),
        offset=offset,
        max_results=max_results,
    )


def get_note(guid: str) -> NoteMetadata:
    """Get note metadata."""
    client = get_client()
    note = client.get_note(guid)
    if _is_private(note.tagGuids or []):
        raise PrivateNoteError(guid)
    return NoteMetadata.from_thrift(note)


def get_note_content(guid: str) -> NoteContent:
    """Get full note content as Markdown."""
    client = get_client()
    note = client.get_note(guid)
    if _is_private(note.tagGuids or []):
        raise PrivateNoteError(guid)
    content = client.get_note_content(guid)
    if note.guid is None:
        raise ValueError("Note returned without GUID")
    return NoteContent(
        guid=note.guid,
        title=note.title or "Untitled",
        content=content,
    )


def list_notebooks() -> list[NotebookInfo]:
    """List all notebooks."""
    client = get_client()
    notebooks = client.list_notebooks()
    return [
        NotebookInfo(guid=nb.guid, name=nb.name, stack=nb.stack)
        for nb in notebooks
        if nb.guid is not None and nb.name is not None
    ]


def list_tags() -> list[TagInfo]:
    """List all tags."""
    client = get_client()
    tags = client.list_tags()
    return [
        TagInfo(guid=t.guid, name=t.name)
        for t in tags
        if t.guid is not None and t.name is not None and t.name.lower() != "private"
    ]


# --- Write operations ---


def create_note(
    title: str,
    content: str,
    notebook_name: str = "",
    tags: list[str] | None = None,
) -> CreatedNote:
    """Create a new note with Markdown content."""
    client = get_client()

    notebook_guid = None
    if notebook_name:
        notebook_guid = resolve_notebook_guid(client, notebook_name)

    note = client.create_note(
        title=title,
        markdown=content,
        notebook_guid=notebook_guid,
        tag_names=tags,
    )
    if note.guid is None or note.title is None:
        raise ValueError("Created note missing GUID or title")
    return CreatedNote(
        guid=note.guid,
        title=note.title,
        notebook_guid=note.notebookGuid,
    )


def tag_note(guid: str, tags: list[str]) -> NoteMetadata:
    """Add tags to an existing note."""
    if any(t.lower() == "private" for t in tags):
        raise PrivateNoteError("Cannot add 'private' tag")
    client = get_client()
    existing = client.get_note(guid)
    if _is_private(existing.tagGuids or []):
        raise PrivateNoteError(guid)
    note = client.tag_note(guid, tags)
    return NoteMetadata.from_thrift(note)


def untag_note(guid: str, tags: list[str]) -> NoteMetadata:
    """Remove tags from an existing note."""
    if any(t.lower() == "private" for t in tags):
        raise PrivateNoteError("Cannot remove 'private' tag")
    client = get_client()
    existing = client.get_note(guid)
    if _is_private(existing.tagGuids or []):
        raise PrivateNoteError(guid)
    note = client.untag_note(guid, tags)
    return NoteMetadata.from_thrift(note)


def move_note(guid: str, notebook_name: str) -> NoteMetadata:
    """Move a note to a different notebook."""
    client = get_client()
    notebook_guid = resolve_notebook_guid(client, notebook_name)
    existing = client.get_note(guid)
    if _is_private(existing.tagGuids or []):
        raise PrivateNoteError(guid)
    note = client.move_note(guid, notebook_guid)
    return NoteMetadata.from_thrift(note)


# --- Write queue ---

_WRITE_DISPATCHER: dict[str, Callable[..., Any]] = {
    "create_note": create_note,
    "tag_note": tag_note,
    "untag_note": untag_note,
    "move_note": move_note,
}


def enqueue_write(operation: str, **params: Any) -> None:
    """Persist a write operation for later execution."""
    from evernote_client.client.queue import OperationQueue

    OperationQueue(settings.queue_path).put(operation, **params)


def pending_write_count() -> int:
    """Return the number of pending queued write operations."""
    from evernote_client.client.queue import OperationQueue

    return OperationQueue(settings.queue_path).size()


def drain_pending_writes() -> int:
    """Process all queued write operations. Returns count processed."""
    from evernote_client.client.queue import OperationQueue

    queue = OperationQueue(settings.queue_path)
    if queue.is_empty():
        return 0
    results = queue.process_all(_WRITE_DISPATCHER)
    return len(results)

File: src/service.ts
import { mkdir, writeFile } from 'node:fs/promises';
import { z } from 'zod';
import { getToken } from './auth/oauth.js';
import { EvernoteClient, PRIVATE_TAG_NAME } from './client/evernote-client.js';
import { OperationQueue } from './client/queue.js';
import { settings } from './config.js';
import type { ResourceInfo } from './enml/types.js';
import { PrivateNoteError } from './errors.js';
import { logger } from './logger.js';
import { safeJoin } from './path-utils.js';
import {
  type CreatedNote,
  type NotebookInfo,
  type NoteContent,
  type NoteMetadata,
  noteMetadataFromThrift,
  type SearchResult,
  type TagInfo,
} from './models.js';

let _client: EvernoteClient | null = null;

export async function getClient(): Promise<EvernoteClient> {
  if (!_client) {
    const token = await getToken(settings);
    _client = new EvernoteClient(token);
  }
  return _client;
}

export function resetClient(): void {
  _client = null;
}

async function isPrivate(tagGuids: string[]): Promise<boolean> {
  const client = await getClient();
  const guid = await client.getPrivateTagGuid();
  return guid !== null && tagGuids.includes(guid);
}

async function assertNotPrivate(
  noteGuid: string,
  tagGuids: string[]
): Promise<void> {
  if (await isPrivate(tagGuids)) {
    throw new PrivateNoteError(noteGuid);
  }
}

async function resolveNotebookGuid(
  client: EvernoteClient,
  name: string
): Promise<string> {
  const notebooks = await client.listNotebooks();
  for (const nb of notebooks) {
    if (nb.name === name && nb.guid) {
      return nb.guid;
    }
  }
  throw new Error(`Notebook not found: ${name}`);
}

async function resolveNotebookGuidIfProvided(
  client: EvernoteClient,
  name: string
): Promise<string | null> {
  return name ? resolveNotebookGuid(client, name) : null;
}

// --- Read operations ---

export async function searchNotes(
  query = '',
  notebookName = '',
  tags?: string[] | null,
  maxResults = 20,
  offset = 0
): Promise<SearchResult> {
  const client = await getClient();
  maxResults = Math.min(maxResults, 100);

  const notebookGuid = await resolveNotebookGuidIfProvided(
    client,
    notebookName
  );

  const safeTags = tags?.filter((t) => t.toLowerCase() !== PRIVATE_TAG_NAME);
  const tagFilter = safeTags?.length ? safeTags : null;

  const result = await client.searchNotes(
    query,
    notebookGuid,
    tagFilter,
    maxResults,
    offset
  );

  const allNotes = (result.notes ?? []).map(noteMetadataFromThrift);
  const notes: NoteMetadata[] = [];
  for (const n of allNotes) {
    if (!(await isPrivate(n.tagGuids))) {
      notes.push(n);
    }
  }
  const filtered = allNotes.length - notes.length;

  // NOTE: The total is approximate — we only know how many private notes were
  // on *this* page. Other pages may contain different counts of private notes.
  return {
    notes,
    total: Math.max(0, (result.totalNotes ?? 0) - filtered),
    offset,
    maxResults,
  };
}

export async function getNote(guid: string): Promise<NoteMetadata> {
  const client = await getClient();
  const note = await client.getNote(guid);
  await assertNotPrivate(guid, note.tagGuids ?? []);
  return noteMetadataFromThrift(note);
}

export async function getNoteContent(
  guid: string,
  options: { resourceDir?: string } = {}
): Promise<NoteContent> {
  const client = await getClient();
  const note = await client.getNote(guid);
  await assertNotPrivate(guid, note.tagGuids ?? []);
  if (!note.guid) throw new Error('Note returned without GUID');

  let content = await client.getNoteContent(guid);

  if (options.resourceDir) {
    const resources = await client.getNoteResources(guid);
    await mkdir(options.resourceDir, { recursive: true });
    for (const r of resources) {
      if (!r.data || !r.filename) continue;
      const target = safeJoin(options.resourceDir, r.filename);
      await writeFile(target, r.data);
      content = content.replaceAll(`evernote-resource:${r.hashHex}`, target);
    }
  }

  return {
    guid: note.guid,
    title: note.title ?? 'Untitled',
    content,
  };
}

export async function listNotebooks(): Promise<NotebookInfo[]> {
  const client = await getClient();
  const notebooks = await client.listNotebooks();
  return notebooks.flatMap((nb) =>
    nb.guid != null && nb.name != null
      ? [{ guid: nb.guid, name: nb.name, stack: nb.stack ?? null }]
      : []
  );
}

export async function listTags(): Promise<TagInfo[]> {
  const client = await getClient();
  const tags = await client.listTags();
  return tags.flatMap((t) =>
    t.guid != null &&
    t.name != null &&
    String(t.name).toLowerCase() !== PRIVATE_TAG_NAME
      ? [{ guid: t.guid, name: t.name }]
      : []
  );
}

// --- Write operations ---

export async function createNote(
  title: string,
  content: string,
  notebookName = '',
  tags?: string[] | null,
  sourceNoteGuid?: string
): Promise<CreatedNote> {
  const client = await getClient();

  const notebookGuid = await resolveNotebookGuidIfProvided(
    client,
    notebookName
  );

  let existingResources: ResourceInfo[] | undefined;
  if (sourceNoteGuid) {
    try {
      existingResources = await client.getNoteResources(sourceNoteGuid);
    } catch (err) {
      logger.warn(
        { err, sourceNoteGuid },
        'Could not fetch source note resources — images will be skipped'
      );
    }
  }

  const note = await client.createNote(
    title,
    content,
    notebookGuid,
    tags && tags.length > 0 ? tags : null,
    existingResources
  );

  if (!note.guid || !note.title) {
    throw new Error('Created note missing GUID or title');
  }

  return {
    guid: note.guid,
    title: note.title,
    notebookGuid: note.notebookGuid ?? null,
  };
}

export async function copyNote(
  sourceGuid: string,
  newTitle: string,
  notebookName = ''
): Promise<CreatedNote> {
  const client = await getClient();

  let toNotebookGuid: string;
  if (notebookName) {
    toNotebookGuid = await resolveNotebookGuid(client, notebookName);
  } else {
    const src = await client.getNote(sourceGuid);
    if (!src.notebookGuid) throw new Error('Source note has no notebook GUID');
    toNotebookGuid = src.notebookGuid;
  }

  const note = await client.copyNote(sourceGuid, newTitle, toNotebookGuid);
  if (!note.guid || !note.title)
    throw new Error('Copied note missing GUID or title');
  return {
    guid: note.guid,
    title: note.title,
    notebookGuid: note.notebookGuid ?? null,
  };
}

async function modifyNoteTags(
  guid: string,
  tags: string[],
  op: 'add' | 'remove'
): Promise<NoteMetadata> {
  if (tags.some((t) => t.toLowerCase() === PRIVATE_TAG_NAME)) {
    const verb = op === 'add' ? 'add' : 'remove';
    throw new PrivateNoteError(`Cannot ${verb} '${PRIVATE_TAG_NAME}' tag`);
  }
  const client = await getClient();
  const existing = await client.getNote(guid);
  await assertNotPrivate(guid, existing.tagGuids ?? []);
  const note =
    op === 'add'
      ? await client.tagNote(guid, tags)
      : await client.untagNote(guid, tags);
  return noteMetadataFromThrift(note);
}

export function tagNote(guid: string, tags: string[]): Promise<NoteMetadata> {
  return modifyNoteTags(guid, tags, 'add');
}

export function untagNote(guid: string, tags: string[]): Promise<NoteMetadata> {
  return modifyNoteTags(guid, tags, 'remove');
}

export async function moveNote(
  guid: string,
  notebookName: string
): Promise<NoteMetadata> {
  const client = await getClient();
  const notebookGuid = await resolveNotebookGuid(client, notebookName);
  const existing = await client.getNote(guid);
  await assertNotPrivate(guid, existing.tagGuids ?? []);
  const note = await client.moveNote(guid, notebookGuid);
  return noteMetadataFromThrift(note);
}

// --- Write queue ---

type WriteDispatcher = Record<
  string,
  (params: Record<string, unknown>) => Promise<unknown>
>;

const CreateNoteParams = z.object({
  title: z.string(),
  content: z.string(),
  notebook_name: z.string().optional().default(''),
  tags: z.array(z.string()).nullable().optional().default(null),
  source_note_guid: z.string().optional(),
});

const GuidTagsParams = z.object({
  guid: z.string(),
  tags: z.array(z.string()),
});

const MoveNoteParams = z.object({
  guid: z.string(),
  notebook_name: z.string(),
});

const WRITE_DISPATCHER: WriteDispatcher = {
  create_note: (p) => {
    const v = CreateNoteParams.parse(p);
    return createNote(
      v.title,
      v.content,
      v.notebook_name,
      v.tags,
      v.source_note_guid
    );
  },
  tag_note: (p) => {
    const v = GuidTagsParams.parse(p);
    return tagNote(v.guid, v.tags);
  },
  untag_note: (p) => {
    const v = GuidTagsParams.parse(p);
    return untagNote(v.guid, v.tags);
  },
  move_note: (p) => {
    const v = MoveNoteParams.parse(p);
    return moveNote(v.guid, v.notebook_name);
  },
};

export function enqueueWrite(
  operation: string,
  params: Record<string, unknown>
): void {
  new OperationQueue(settings.queuePath).put(operation, params);
}

export function pendingWriteCount(): number {
  return new OperationQueue(settings.queuePath).size();
}

export async function drainPendingWrites(): Promise<number> {
  const queue = new OperationQueue(settings.queuePath);
  if (queue.isEmpty()) return 0;
  const results = await queue.processAll(WRITE_DISPATCHER);
  return results.length;
}
